### PR TITLE
Synthesis of non-recursive programs

### PR DESCRIPTION
**Pull request by hzzv**
Brief intro:
The main idea of synthesis is that we use answer literals to track program fragments throughout the proof search.
If we derive a clause `C \/ ans(r)`, then `if ~C then r` is a conditional program. Therefore, we record the program `r` with condition `~C`, and replace the clause `C \/ ans(r)` in the search space with `C` (so that we guide the proof search towards reasoning about the cases when `C` holds).
After concluding the proof search, we construct a program `if ~C_1 then r_1 else if ~C_2 then r_2 else ... if C_{n-1} then r_{n-1} else r_n` from the recorded programs `r_i` with conditions `C_i`.

This PR introduces the following changes:
- option `--question_answering synthesis` which turns on synthesis using answer literals
- answer literals are ignored by LiteralIndex, DemodulationSubtermIndexImpl, and ForwardDemodulation
- symbols in SMTLIB2 syntax can be marked as "uncomputable" using `(set-option :uncomputable (symbol1 ... symbolN))`
- modifying inference rules (binary resolution, equality factoring, equality resolution, factoring, superposition, URR) such that they do not introduce uncomputable symbols into the answer literal
- modifying binary resolution, superposition and URR such that if they have multiple premises with answer literal, the resulting clause will only have one answer literal using if-then-else (for URR this requires adding two new indices)
- adding SynthesisManager class derived from AnswerLiteralManager, which:
(1) adds answer literals for specifications in form `~ !x. ?y. F[x,y]`,
(2) records the correspondence between Skolem constants and variables `x` from the specification,
(3) records program `r` with condition `~C` when a clause `C \/ ans(r)` is derived where `C` is computable,
(4) after concluding the proof search constructs the final program as `if ~C_1 then r_1 else if ~C_2 then r_2 else ... if C_{n-1} then r_{n-1} else r_n` from all the recorded programs with conditions, where the Skolem constants corresponding to `x` are replaced by `x`
- AVATAR integration with answer literals (allowing splitting but only for computable ground clauses without answer literals)
- adding two new inferences:
ANSWER_LITERAL_REMOVAL which replaces clause `C \/ ans(r)` with `C` when `C` is computable (after recording the program `r` as explained above); and 
AVATAR_ASSERTION_REINTRODUCTION, which replaces clause `C \/ ans(r) <- A1,...,An` with `C \/ ans(r) \/ ~A1 \/ ... \/ ~An`
- the if-then-else used within answer literals is a newly introduced function symbol with only literals as conditions (so that clauses during proof search do not contain special terms), while the program generated after the proof search concludes uses the special-term if-then-else constructor (so that formulas can be used as conditions)
- the calculus using the rules with computability restriction is incomplete (since some inferences are blocked), so saturation does not entail satisfiability

### CODE REVIEW

**Comment by hzzv** [comment]
There is one outstanding question I haven't resolved yet: for clauses, literals and terms we need to compute whether they are computable or not. Should we recompute this information every time it's needed, or add it as a member bool variable for Clause and Term? What do you think?

For clauses we need this information when we're deciding whether to split and whether to use it for ANSWER_LITERAL_REMOVAL.
For literals/terms we check this if we need to check the computability of a clause containing the literal/term.
(Right now the implementation caches this for clauses but not for terms/literals due to me being lazy and not bothering with caching it for terms when I'm not sure it's the best way.)

**Comment by MichaelRawson** [comment]
Cool! As usual I'm not reviewing the question-answering logic, only the interface with the existing Vampire. Some initial questions without reading the code or thinking about the approach too much:
1. Does this subsume existing answer-literal/question-answering approaches? Seems like it might - if so, these should be removed or upgraded.
2. It's intended and useful for answer literals to participate in general-purpose inferences, right? If so: why no demodulation? If not: would it be easier just to _never_ index answer literals?
3. Generally I'd lean towards "positive" input features i.e. marking symbols as computable explicitly, rather than uncomputable. How much effort is it to flip the logic?
4. "do not introduce uncomputable symbols into the answer literal" - would it be easier to allow the superposition anyhow and then delete clauses that have uncomputable symbols in answer literals?
5. " if they have multiple premises with answer literal, the resulting clause will only have one answer literal using if-then-else" - could this be done outside the inference, or does it depend on some specifics of the inferences?
6. What's the intuition behind `AVATAR_ASSERTION_REINTRODUCTION`?

> Should we recompute this information every time it's needed, or add it as a member bool variable for Clause and Term?

Generally the philosophy seems to be that Vampire caches this stuff to prevent real or imagined performance problems when somebody inevitably calls `computable()` on a term in a loop. We can go hunting for a spare bit in `Term *` or it might be OK to keep a hashmap somewhere.

**Comment by hzzv** [comment]
> 1. Does this subsume existing answer-literal/question-answering approaches? Seems like it might - if so, these should be removed or upgraded.

Good question. It can either preserve the previous behavior of question answering (generating disjuntive answers), or it can replace it (generating definite answers using if-then-else). @selig which option would be preferable?

> 2. It's intended and useful for answer literals to participate in general-purpose inferences, right? If so: why no demodulation? If not: would it be easier just to _never_ index answer literals?

The intention was to completely exclude answer literals from all inferences.
The answer literals are never indexed by LiteralIndex, but to exclude them from demodulation, we also need to modify both ForwardDemodulation and DemodulationSubtermIndexImpl.

> 3. Generally I'd lean towards "positive" input features i.e. marking symbols as computable explicitly, rather than uncomputable. How much effort is it to flip the logic?

Flipping the logic would be very inconvenient for the user.
Usually a large majority symbols is computable. We don't want to burden the user by having to annotate every symbol used in the input. Further, all symbols being considered computable by default agrees with interpreted symbols being always computable.

> 4. "do not introduce uncomputable symbols into the answer literal" - would it be easier to allow the superposition anyhow and then delete clauses that have uncomputable symbols in answer literals?

The implementation would probably be simpler. However, I don't have good intuition on how would this impact the performance, since it would lead to generating potentially many clauses that are immediately deleted. I'd be happy to discuss this further (maybe in person?).

> 5. " if they have multiple premises with answer literal, the resulting clause will only have one answer literal using if-then-else" - could this be done outside the inference, or does it depend on some specifics of the inferences?

Depends on the specifics of the inference, the `if-then-else` condition uses literals from the premises. E.g. for binary resolution, if we resolve `L \/ ans(r1)` against `~L \/ ans(r2)`, we obtain `ans(if L then r2 else r1)`.

> 6. What's the intuition behind `AVATAR_ASSERTION_REINTRODUCTION`?

We cannot split a clause with answer literal - hence if we obtain a clause with avatar assertions that has an answer literal, we have to de-split it :)
Also, since only ground and computable clauses are allowed to be split, `AVATAR_ASSERTION_REINTRODUCTION` is always immediately followed by `ANSWER_LITERAL_REMOVAL` (and can be then split further, since the replacement does not contain an answer literal anymore).

> Generally the philosophy seems to be that Vampire caches this stuff to prevent real or imagined performance problems when somebody inevitably calls `computable()` on a term in a loop. We can go hunting for a spare bit in `Term *` or it might be OK to keep a hashmap somewhere.

Okay! Which of the two options should I look into?

**Comment by quickbeam123** [comment]
Hey, nice work! Starting to look... 

As I merged Marton's PR few minutes ago, this no longer compiles (some issues in `SMTLIB2.cpp`) when merged with master.

I am also running some performance check - so that we have a better idea of how expensive it is to look for an answer literal in every clause we do BinaryResolution on.

Will come back with more comments hopefully :) next week!

**Comment by MichaelRawson** [comment]
> Okay! Which of the two options should I look into?

I think this only needs to work for `Term`, not `TermList`, right? If so things are a lot less constrained. You can either add a new `bool` field with the other metadata and take the hit (which might be zero), or some attempt at micro-optimisation has been made in the past via bitfields. I'd expect `_functor` could be `_functor: 31` and use the spare bit for this purpose.

File: Shell/AnswerExtractor.cpp
Comment (MichaelRawson) [inline Shell/AnswerExtractor.cpp:41]: ...uh. C++ name resolution makes my head hurt, can you explain this anonymous namespace for me?

File: Shell/Skolem.cpp
Comment (MichaelRawson) [inline Shell/Skolem.cpp:99]: Should this logic be duplicated in NewCNF? I think it has its own Skolem routine. Also potentially definition introduction?

File: Kernel/Clause.hpp
Comment (MichaelRawson) [inline Kernel/Clause.hpp:415]: Could this be an `unsigned` (possibly with fewer bits) indicating the position of the literal in the clause?

**Review by MichaelRawson — COMMENTED** [review]
Hi @hzzv, code looks great - mostly conceptual comments. Let's discuss today. :-)

**Comment by MichaelRawson** [comment]
I believe this is in the "good" state of PRs where we are waiting for @hzzv to address comments and merge conflicts, so nothing needs to be done. :-)

There was some discussion about whether we want to have "extras" in mainline Vampire or have them as satellite branches until they coalesce enough to be manageable in mainline. Generally I'm inclined to the "satellite" solution, but in this case it seems to be a strict improvement on the question-answering functionality that already exists and people rely on, so I won't lose too much sleep about (eventually) merging it.

**Comment by quickbeam123** [comment]
I also vaguely remember I wanted to comment on the "unprotected" (i.e. uncondiotional) calls (e.g. in Resolution.cpp) to functions only needed when question answering, like the search for an answer literal. I suppose you (Micheal + Petra) might have discussed this already. In any case, I will be happy to explain this in more detail when the branch looks "otherwise ready for merge".

**Comment by hzzv** [comment]
Yes, waiting for me is indeed the state. I have it in my todo list, but I might only get around to it after the internship.

**Comment by MichaelRawson** [comment]
Thanks for confirmation, Petra - this is not a "hurry up", I'm just doing some cleanup with open merges as you probably noticed. Hope the internship is going well!

File: Inferences/URResolution.cpp
Comment (hzzv) [reply Inferences/URResolution.cpp]: Done.

File: Kernel/Clause.cpp
Comment (hzzv) [reply Kernel/Clause.cpp:198]: Indeed.

File: Kernel/Term.cpp
Comment (hzzv) [reply Kernel/Term.cpp:863]: Done.

File: Kernel/Term.cpp
Comment (hzzv) [reply Kernel/Term.cpp:1141]: Yes. I think we discussed this back in May, but let me know if you need more info :)

File: Kernel/Term.cpp
Comment (hzzv) [reply Kernel/Term.cpp]: I removed the `isEquality`.
Is the optimization worth the additional dependencies that would have to be included in the header file? (Environment, Signature, possibly more...?)

File: Kernel/TermTransformer.hpp
Comment (hzzv) [reply Kernel/TermTransformer.hpp:47]: Apparently not :)

File: Saturation/SaturationAlgorithm.cpp
Comment (hzzv) [reply Saturation/SaturationAlgorithm.cpp]: Done.

File: Saturation/Splitter.hpp
Comment (hzzv) [reply Saturation/Splitter.hpp:223]: Done.

File: Shell/AnswerExtractor.cpp
Comment (hzzv) [reply Shell/AnswerExtractor.cpp]: As discussed in person, I removed it.

**Review by hzzv — COMMENTED** [review]
I resolved most of the comments.

What still needs to be done is (1) check if we can subsume the question answering logic using synthesis, (2) change the logic of individual inferences so that they don't check if an answer literal is uncomputable, and instead just derive new clauses, and then use InvalidAnswerLiteralRemoval to delete clauses with invalid answer literals (I added a class for this, Daneshvar is working on implementation), (3) duplicate the logic from Shell/Skolem.cpp in NewCNF.

@MichaelRawson - is it correct to derive the deleting inference for (2) from ImmediateSimplificationEngine? Or should we rather derive it from ForwardSimplificationEngine (or some other *Engine)? The deletion only depends on the clause itself, however, it looks to me like from all ImmediateSimplificationEngines, only the first one that applies is actually executed for each clause: https://github.com/vprover/vampire/blob/master/Inferences/InferenceEngine.cpp#L101

**Comment by MichaelRawson** [comment]
> I resolved most of the comments.

Thanks! Lmk when ready for review.

> @MichaelRawson - is it correct to derive the deleting inference for (2) from ImmediateSimplificationEngine? Or should we rather derive it from ForwardSimplificationEngine (or some other *Engine)? The deletion only depends on the clause itself, however, it looks to me like from all ImmediateSimplificationEngines, only the first one that applies is actually executed for each clause: https://github.com/vprover/vampire/blob/master/Inferences/InferenceEngine.cpp#L101

I think so! My understanding is immediate simplifications only depend on the clause, forward simplifications depend on the active set - so this should be an immediate simplification. Once a clause is deleted, it's deleted - *why* it was deleted doesn't matter to the rest of the saturation loop. Did I miss something?

**Comment by hzzv** [comment]
> My understanding is immediate simplifications only depend on the clause, forward simplifications depend on the active set - so this should be an immediate simplification. Once a clause is deleted, it's deleted - _why_ it was deleted doesn't matter to the rest of the saturation loop. Did I miss something?

Ah, I didn't realize the replacement clause is then added as new, and goes through the simplification again. Makes sense, thanks!

There are still a couple of outstanding tasks which we're working on. I have some questions about the skolemization in NewCNF - but if possible, it would be easier to discuss that in person.

File: Kernel/Term.cpp
Comment (hzzv) [reply Kernel/Term.cpp]: As discussed offline, the optimization is not worth it.

File: Shell/Skolem.cpp
Comment (hzzv) [reply Shell/Skolem.cpp]: Done as discussed offline: we now skolemize the input variables from within SynthesisManager.

File: Kernel/Clause.hpp
Comment (hzzv) [reply Kernel/Clause.hpp:405]: As discussed offline, we will keep it as a pointer, since not all clauses have an answer literal.

**Comment by hzzv** [comment]
The PR is ready for a review, @MichaelRawson!

The large changes since the last time you took a look at the code:
1. We now skolemize the input variables from inside of SynthesisManager. This allows us to mark those skolems as computable (and store their binding to variables), while all other skolems are uncomputable.
2. The inference rules now do not check answer literal computability. Rather, clauses with uncomputable answer literals are deleted in immediate-simplification step.
3. We also added parsing of uncomputable annotation in TPTP.

File: Shell/AnswerExtractor.hpp
Comment (MichaelRawson) [inline Shell/AnswerExtractor.hpp:105]: This isn't polymorphism-safe, and introduces one symbol per sort in the many-sorted case. Could this be a single polymorphic symbol `$ite: !>[X: $tType]: ($o * X * X) > X` instead?

**Review by MichaelRawson — APPROVED** [review]
This is a big improvement! I'm a happy bunny. I have a few more minor questions from another scan of the code, but looks fine to me. Well done to both of you.

Before merge, could the Z3 submodule get set back to whatever version it is on master please? I'm not sure why this keeps getting set to something different, lots of people get tripped up.

**Comment by MichaelRawson** [comment]
@quickbeam123 - once @hzzv and/or @daneshvar-amrollahi answer/fix the last few comments I have, I'd be inclined to merge this.

Generally I think we agree that experimental stuff stays in branches, *but* this one is quite small and as far as we can tell supersedes (some of) the other question-answering modes, which would allow us to remove them in a future PR. I trust @hzzv to at least remind me to do that, if not do the surgery herself. :-) Keen to hear your thoughts too if you have the time.

File: Inferences/InvalidAnswerLiteralRemoval.cpp
Comment (daneshvar-amrollahi) [reply Inferences/InvalidAnswerLiteralRemoval.cpp:12]: Added a comment in the header file, right before the class declaration.

File: Inferences/InvalidAnswerLiteralRemoval.hpp
Comment (daneshvar-amrollahi) [reply Inferences/InvalidAnswerLiteralRemoval.hpp]: Removed.

File: Inferences/InterpretedEvaluation.cpp
Comment (hzzv) [reply Inferences/InterpretedEvaluation.cpp]: Actually we do want them. Removed, thanks!

File: Shell/AnswerExtractor.hpp
Comment (MichaelRawson) [reply Shell/AnswerExtractor.hpp]: After discussion: this might be possible, but polymorphism and synthesis needs further investigation in a separate PR anyway. Keep it as-is for now.

**Comment by MichaelRawson** [comment]
Everything resolved from my point of view. Thanks @hzzv and @daneshvar-amrollahi!

**Comment by quickbeam123** [comment]
Oh, by the way, @hzzv , could you please quickly check whether being now an expert on QA wouldn't make incorporating also https://github.com/vprover/vampire/pull/108 a trifle for you? (@MichaelRawson : Two PRs for the price of one, doesn't that sound great?)

**Comment by hzzv** [comment]
Based on what I understand from the code of #108, I think those changes are orthogonal to this PR, and therefore I would like to avoid joining the PRs.
Also I'm still waiting on @selig to respond whether he thinks we can remove (parts of) the QA code after merging this. In a way, synthesis supersedes it, but also not completely.

File: Shell/Preprocess.cpp
Comment (hzzv) [reply Shell/Preprocess.cpp]: Done.

File: Parse/TPTP.cpp
Comment (quickbeam123) [inline Parse/TPTP.cpp:4996]: Getting
``Parse/TPTP.cpp:4996:20: warning: ‘color’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 4996 |       sym->addColor(color);``
 here with `gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0`. Could we add some default value at line 4967?

File: Kernel/Clause.cpp
Comment (quickbeam123) [reply Kernel/Clause.cpp:196]: Could I just curiously inquire what kind of bug? 

You mean `proofExtra` storing info for a deleted clause? That means a memory leak? Or something more dangerous?

**Comment by quickbeam123** [comment]
While a performance test is running, I wanted to ask if we could together think of how to make this PR the least impactful on the main execution paths of vampire when it's not doing question answering or synthesis. This basically boils down to the code of the affected inference rules. 

I am, in particular, concerned about the functions `hasAnswerLiteral`/`getAnswerLiteral` which keep scanning the clause's literals (also when outside of question answering modes); this could start making a difference with clauses with tens of literals or more. Perhaps we could guard calls to these with something along the lines of `options.qa != off`? This could perhaps also help with identifying the pieces of code there which are qa-specific and can be skipped when looking for the main logic of each inference's code.

By the way, this PR seems to violate vampire's implicit coding standard of always putting if-statement body into braces. E.g.,

`if (neg) resolved = Literal::complementaryLiteral(resolved);`

I have spent too much time looking at all the other code not to get confused by this for a split of a second. However, let's discuss this, if you care about your coding freedom more! ;)

File: Kernel/Clause.cpp
Comment (hzzv) [reply Kernel/Clause.cpp:196]: Since I wrote this code already quite some time ago, I don't remember exactly what the bug was - but I think it was about storing info for a deleted clause, then allocating new clause in the same memory slot, and hence having wrong info for the new clause.

File: Inferences/Factoring.cpp
Comment (quickbeam123) [inline Inferences/Factoring.cpp]: Thank you, @hzzv , for immediately jumping to action. However, should if-condition in Factoring really disappear completely?

File: Kernel/Clause.cpp
Comment (MichaelRawson) [reply Kernel/Clause.cpp:196]: Already does! https://github.com/vprover/vampire/blob/master/Lib/Hash.hpp#L145 :-)

File: Inferences/Factoring.cpp
Comment (hzzv) [reply Inferences/Factoring.cpp]: Actually it can. We now have the `InvalidAnswerLiteralRemoval` inference which removes clauses with uncomptuable arguments in answer literals. This even further decreases the side effects on the main flow when synthesis isn't on.

Comment (quickbeam123) [reply Inferences/Factoring.cpp]: Perfect!

**Comment by quickbeam123** [comment]
Thank you again, @hzzv, for the quick updates. I restarted the perf-check. I think the last bit that we will potentially discuss is URR, but let's see how the test goes...

**Comment by hzzv** [comment]
Thanks @quickbeam123!

As you've seen, I added checks if synthesis is on for all calls of {has,get}AnswerLiteral in the main path.
I also changed all the `if (cond) <code>` to `if (cond) { <code> }`.
I'm all up for coding style consistency - but it would be great to know about it beforehand. (This is exactly the reason why I pushed for a code style guide at the last Vampire dev meeting.)

I now noticed there is one more thing for which @selig might offer some input: do we want to have computability checks for question answering as well? E.g., is a skolem constant coming from some skolemized axiom a valid answer? I'd say it should not be, but before this PR it was allowed.
I might do a couple minor changes related to this later today. I'll let you all know when I'm done.

**Comment by quickbeam123** [comment]
While this is getting finalised...

> E.g., is a skolem constant coming from some skolemized axiom a valid answer?

I think that's fine. The way I would read the answer would be "The object you guaranteed to exist in axiom XYZ, gives rise to an answer to the question you posed." (This is only a bit problematic as the connection between the skolem and the the axiom is lost in the symbol and needs to be recovered from a preprocessing log or from a proof, but otherwise, I think, legit.)

Some partial experiment results. First for a strategy involving URR:
master        8030.6 +/- 10.0
synth6922  8015.7  +/- 10.0
synth6925  8026.7 +/- 10.4
So your changes from today morning helped quite a lot here! (We could still think how to get the overhead even closer to zero.)

Plain discount experiments are still running.

**Comment by hzzv** [comment]
Ok, I removed computability check for question answering.

Let me know how is the overhead looking overall when the plain discount experiments finish.

**Comment by quickbeam123** [comment]
With plain discount, adding the if's appernetly made things a bit worse:
master        8745.0 +/- 11.3
synth6922 8745.7 +/- 11.2
synth6925 8742.1 +/- 11.4
Actually, looking for the answerLit and failing there was actually a bit faster, on average, than not doing it all (as in master), which I find hard to explain...

**Comment by hzzv** [comment]
That is indeed surprising, I also don't have an explanation... Is there anything we can do on our side about it?

**Comment by MichaelRawson** [comment]
>Actually, looking for the answerLit and failing there was actually a bit faster, on average, than not doing it all (as in master), which I find hard to explain...

Is it possible to find problems that were most affected by this in your harness? I also don't have an explanation, but I wouldn't be _so_ surprised, could be e.g. enabled compiler optimization, cache pre-heat.

**Comment by quickbeam123** [comment]
I was also thinking something along the lines of cache pre-heat, but it's really hard to pinpoint. While the overall average might not by lying (much), when I single out problems with extreme "probability of getting solved" difference, the result is not reproducible on them - they seem to simply have gotten lucky during the eval with their random seeds... By the way, the fact there is this tiny improvement with synth6922 over master with "-i 10K" does not mean it would persist for longer lasting runs. So I wouldn't spend too much time on it anymore.

However, we could still spend some extra time (a small epsilon, so not too much) thinking whether we could possibly save the (master 8030.6 +/- 10.0) - (synth6925 8026.7 +/- 10.4) many problems getting lost in the new URR by having a whole new class (sharing most of its code with the old one via inheritance? / templating?) that would only get loaded instead of the good old URR implementation when needed (i.e., under the synthesis option). I predict the overhead would stop being measurable (outside of synthesis benchmarks, of course). Whether the gain is worth the coding (now) and maintenance (in the future) is something I would like to quickly discuss with you.

**Comment by hzzv** [comment]
My priority is getting this merged into master - so I can take care of creating the new URR-class supporting answer literals.
However, I might not be around for the maintenance much longer, so that decision is ultimately up to you.

**Comment by MichaelRawson** [comment]
Not against (i) the new URR creation or (ii) having a chat about it.

Buuuut - seems like the current URR is just doing something like
```c
if(we_might_have_answer_literals)
  do_complex_logic();
else
  do_regular_logic();
```

and if `we_might_have_answer_literals` is reasonably (branch, prefetch) predictable then it should almost not show up at all. So this is a bit surprising to me. Guess the only way is to try it and see.

**Comment by hzzv** [comment]
Martin is right with pointing out how many checks the new code adds... let me see, maybe I can optimize it a bit without introducing a whole new class.

**Comment by quickbeam123** [comment]
@hzzv, please do not spend too much time on it. Unless you see some elegant solution, I would say it's not worth it this time. (4 problems under urr, which is not a key inference). I mainly just wanted this to be some food for thought and an example of the decisions we are facing when merging new features.

**Comment by hzzv** [comment]
I moved some of the checks for synthesis/answer literals in URR around. Maybe it will help a bit. Otherwise I don't see any elegant solution.

**Comment by quickbeam123** [comment]
Excellent, thank you! I will evaluate one more time, but unless you made a mistake by accident, let's call it a ready PR and merge later today.

**Comment by hzzv** [comment]
Thanks! I checked that we still solve all the synthesis benchmarks, so the chances for a mistake are low :)

**Comment by quickbeam123** [comment]
For some reason, your changes - although looking quite reasonable - unfortunately didn't help:
URR
master 8030.6 +/- 10.0
synth6922 8015.7 +/- 10.0
synth6925 8026.7 +/- 10.4
synth6927 8019.9 +/- 10.3

**Comment by quickbeam123** [comment]
I suggest we revert the last commit (I can do this) and merge.

**Comment by hzzv** [comment]
Okay - yes, please go ahead. Thank you :)

**Comment by quickbeam123** [comment]
Thank you too! Good job!

**Pull request merged**
